### PR TITLE
BUG: ensure consistent behavior of Index.is_all_dates (#19204)

### LINKED ISSUE #19204: "is_all_dates" inconsistent across datetimelike types
xref #19187 

#### Code Sample, a copy-pastable example if possible

```
In [2]: pd.DatetimeIndex(['2000-02-01', '2000-02-02']).is_all_dates
Out[2]: True

In [3]: pd.DatetimeIndex(['2000-02-01', '2000-02-02']).append(pd.Index(['not a date']))[:2].is_all_dates
Out[3]: True

In [4]: pd.PeriodIndex(['2000', '2001'], freq='A').is_all_dates
Out[4]: True

In [5]: pd.PeriodIndex(['2000', '2001'], freq='A').append(pd.Index(['not a date']))[:2].is_all_dates
Out[5]: False

In [6]: pd.TimedeltaIndex([2000, 2001]).append(pd.Index(['not a date']))[:2].is_all_dates
Out[6]: False
```

#### Problem description

``is_all_dates`` is smart for ``object`` indexes but only for ``DatetimeIndex``, which is inconsistent since the other datetimelike indexes are also considered ``is_all_dates``.

#### Expected Output

``Out[5]:`` and ``Out[6]:`` should be the same as ``Out[3]:``.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

</details>

### PR DESCRIPTION
**Pull request by addammy**
- [x] closes #19204
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -1206,6 +1206,16 @@ def test_is_object(self, index, expected):
     def test_is_all_dates(self, index, expected):
         assert index.is_all_dates is expected
 
+    @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: can you just add these onto the above test

**Review by jreback — CHANGES_REQUESTED** [review]
look ok, merge master and ping on green

File: pandas/tests/indexes/test_base.py (hunk 1 above)
Comment (addammy) [reply pandas/tests/indexes/test_base.py]: Do you want to merge those two tests into one or just to extend the list of fixtures in test_is_all_dates?

Comment (addammy) [reply pandas/tests/indexes/test_base.py:1213]: I extended the test above to check if `is_all_dates==True` for all of the date types. This was indeed missing.
But the test I added checks for something different, what was exposed in the bug report: if `is_all_dates` remains consistent for every date type index during manipulation.

**Comment by jbrockmendel** [comment]
@addammy can you rebase

**Comment by addammy** [comment]
> @addammy can you rebase

Done

**Comment by jbrockmendel** [comment]
@addammy can you try to get the CI passing

**Comment by jreback** [comment]
can you merge master; ping on green.

**Comment by jreback** [comment]
closing as stale
if u want to continue pls ping

**Pull request closed without merging**